crates/win-api-wrappers/src/netmgmt.rs

```diff
@@ -60,74 +58,61 @@ pub fn get_local_admin_group_members() -> anyhow::Result<Vec<Sid>> {
-
-    // SAFETY: Verify that all the safety preconditions of from_raw_parts are uphold: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety
-    let group_members_slice = unsafe {
-        std::slice::from_raw_parts(
-            group_members_ptr.0 as *const u8 as *const LOCALGROUP_MEMBERS_INFO_0,
-            entries_read as usize,
         )
     };
 
-    let mut group_members = Vec::<Sid>::with_capacity(group_members_slice.len());
-    for member in group_members_slice {
-        group_members.push(Sid::try_from(member.lgrmi0_sid)?);
+    if ret != NERR_Success {
+        bail!(Error::from_win32(WIN32_ERROR(ret)))
     }
 
-    Ok(group_members)
-}
+    // SAFETY:
+    // - `NetLocalGroupGetMembers` sets `group_members` to a valid pointer on success.
+    // - `group_members` must be freed by `NetApiBufferFree`.
+    // - For level = 0, bufptr will be set to a pointer to a LOCALGROUP_MEMBERS_INFO_0.
+    let group_members = unsafe { NetmgmtMemory::from_raw(group_members.cast::<LOCALGROUP_MEMBERS_INFO_0>()) };
 
-/// RAII wrapper for Network Management memory.
-struct NetMgmtMemory(*mut core::ffi::c_void);
+    // SAFETY:
+    // - `NetLocalGroupGetMembers` returns a pointer valid for `number_of_entries_read` reads of `LOCALGROUP_MEMBERS_INFO_0`.
+    // - `NetLocalGroupGetMembers` is also returning a pointer properly aligned.
+    // - We ensure the memory referenced by the slice is not mutated by shadowing the variable.
+    // - There are never so many entries that `number_of_entries_read * mem::size_of::<LOCALGROUP_MEMBERS_INFO_0>()` overflows `isize`.
+    let group_members = unsafe { group_members.cast_slice(number_of_entries_read as usize) };
 
-impl NetMgmtMemory {
-    // SAFETY: `ptr` must be a valid pointer to memory allocated by Network Management.
-    unsafe fn new(ptr: *mut core::ffi::c_void) -> Self {
-        Self(ptr)
-    }
+    let group_members = group_members
+        .iter()
+        .map(|member| Sid::try_from(member.lgrmi0_sid))
+        .collect::<Result<Vec<Sid>, _>>()?;
+
+    Ok(group_members)
 }
 
-impl Drop for NetMgmtMemory {
-    fn drop(&mut self) {
-        if self.0.is_null() {
-            return;
-        }
+struct NetmgmtFreeMemory;
 
-        // SAFETY: FFI call with no outstanding precondition.
-        unsafe { NetApiBufferFree(Some(self.0)) };
+impl crate::memory::FreeMemory for NetmgmtFreeMemory {
+    /// # Safety
+    ///
+    /// `ptr` is a pointer which must be freed by `NetApiBufferFree`
+    unsafe fn free(ptr: *mut core::ffi::c_void) {
+        // SAFETY: Per invariant on `ptr`, NetApiBufferFree must be called on it for releasing the memory.
+        unsafe { NetApiBufferFree(Some(ptr)) };
     }
 }
 
-impl Default for NetMgmtMemory {
-    fn default() -> Self {
-        Self(std::ptr::null_mut())
-    }
-}
+type NetmgmtMemory<T = core::ffi::c_void> = crate::memory::MemoryWrapper<NetmgmtFreeMemory, T>;
```

Comment:
Rust naming convention:

> In UpperCamelCase, acronyms and contractions of compound words count as one word: use Uuid rather than UUID, Usize rather than USize or Stdin rather than StdIn. 

https://rust-lang.github.io/api-guidelines/naming.html